js browser sdk, initial version

megamk opened the pull request:
This is initial version with basic tests and examples. 
Examples and tests should work with file loaded from https://rawgit.com/megamk/zello-channel-api/js-sdk/sdks/js/dist/zcc.Sdk.js  but you can also build manually

todo: 
- [ ] add more tests
- [x] add documentation for third-party classes (`ZCC.Recorder`, `ZCC.Player.Decoder` and `ZCC.Player.PCMPlayer`) 
- [x] style widget and add functions (volume change, connect/disconnect) and events to widget
- [ ] nodejs support

documentation (`USAGE.md`) is generated using [jsdoc-to-markdown](https://www.npmjs.com/package/jsdoc-to-markdown) still looking for better config options for it. Don't modify this file manually.

megamk commented on sdks/js/package-lock.json:
```diff
@@ -0,0 +1,10472 @@
+{
```
@agatx there was one of your pull requests not so long ago where I mentioned that this file shouldn't be added to git. I was wrong, they actually [recommend to add it to vcs](https://docs.npmjs.com/files/package-lock.json)

vocoded commented on sdks/js/README.md:
````diff
@@ -0,0 +1,50 @@
+No need to install, just clone
+```bash
+> cd src/vendor
+git clone git@github.com:megamk/opus-to-pcm.git
````
We should move these off your personal account, right?

megamk replied:
> We should move these off your personal account, right?

I can fork this into zelloptt, yes.  All three projects were forked because I had to modify them

vocoded commented on sdks/js/src/classes/sdk.js:
```diff
@@ -0,0 +1,201 @@
+  }
+
+  static autoInit() {
+    let autoInitSession = initOptions.session && initOptions.session.autoInit;
```
can use `const` for all these

vocoded commented on sdks/js/src/classes/sdk.js:
```diff
@@ -0,0 +1,201 @@
+    Sdk.session = new library.Session(initOptions.session);
+    let connectCallback = null;
+    let logonCallback = null;
+    if (initOptions.session.onConnect && typeof initOptions.session.onConnect === 'function') {
```
Should be able to drop the first part of this check (typeof will be undefined if missing)

megamk replied:
👍

vocoded commented on sdks/js/src/classes/sdk.js:
```diff
@@ -0,0 +1,201 @@
+      typeof window.WebSocket !== 'function' ||
+      typeof JSON.stringify !== 'function' ||
+      typeof JSON.parse !== 'function' ||
+      typeof Object.assign !== 'function'
```
Probably worth noting browser requirements in the readme (not sure if we're polyfilling this elsewhere)

megamk replied:
will do

vocoded commented on sdks/js/src/classes/session.js:
```diff
@@ -0,0 +1,286 @@
+    let dfd = Promise.defer();
+    this.wsConnection = new WebSocket(this.initialParams.serverUrl);
+    this.wsConnection.binaryType = 'arraybuffer';
+    let connnected = false;
```
typo

vocoded commented on sdks/js/src/classes/session.js:
```diff
@@ -0,0 +1,286 @@
+    params.seq = this.getSeq();
+    params.command = 'stop_stream';
+    let dfd = Promise.defer();
+    let callback = (err, data) => {
```
This definition is the same as the method above.  If we're going to keep it we could factor out

vocoded commented on sdks/js/src/classes/widget/index.js:
```diff
@@ -0,0 +1,344 @@
+  constructor(params) {
+    super();
+    this.options = params;
+    if (!this.options.widget) {
```
Could factor these checks to a validate method similar to the other class

vocoded commented on sdks/js/src/classes/widget/index.js:
```diff
@@ -0,0 +1,344 @@
+    let millis = Math.round((duration % 1000) / 100);
+
+    if (millis >= 10) {
+      millis = 2;
```
why does this reset to 2?

megamk replied:
this, this looks like a bug (will return 00:6.2 for 6999ms), but also I copied this old snippet from zello.com code, so not sure about the origin. Will fix

megamk commented on sdks/js/USAGE.md:
````diff
@@ -0,0 +1,254 @@
+
+Load sdk using `script` tag:
+```html
+<script src="https://zello.com/zcc/0.0.1/zcc.Sdk.js"></script>
````
I wanted libraries classes to be capitalized, because IDE gives warnings about lowercase constructors. With webpack default options it names output files like classes, but we can lowercase during deployment. Will also look into webpack option to see if there's a proper solution.
CDN is also possible, yes.

megamk commented on sdks/js/USAGE.md:
```diff
@@ -0,0 +1,254 @@
+    onConnect: function(err, result) {},
+    onLogon: function(err, result) {}
+  }
+}, function(err) {
```
I added examples with both at the same time, you're right, in real life it makes no sense, so I guess I will split examples into callbacks and promises

megamk commented on sdks/js/USAGE.md:
````diff
@@ -0,0 +1,254 @@
+**Example**  
+```js
+ZCC.Sdk.init({
+  recorder: document.location.protocol.match(/https/),
````
it's not required and `false` by default. But for universal usage (on the same page that can be served both over `http` and `https` I added this as a dynamic param, because trying to init recorder on `http:` will throw an error

megamk commented on sdks/js/USAGE.md:
```diff
@@ -0,0 +1,254 @@
+  });
+  widget.setSession(session);
+})
+.fail(function(err) {
```
thanks, `Q` provides `fail` as an alias to `catch`, will replace with `catch`

megamk commented:
@vocoded @agatx 
Ready for another review round. Restructured to make more similar to iOS. Options to override default components (player/decoder and recorder/encoder). Updated documentation (jsdoc -> html), added JS related documentation into main README.md, updated examples

megamk commented:
will create new PR with no docs for easier review
